asap/em_montage_qc/distorted_montages.py

```diff
@@ -0,0 +1,203 @@
+
+    return jp
+
+
+'''
+def groupId_from_tilespec(ts):
+    return ts.layout.sectionId
+
+def sections_from_resolvedtiles(rts):
+    return list({groupId_from_tilespec(ts) for ts in rts.tilespecs})
+
+def get_z_scales_nopm(z, input_stacks, render):
+    rts = get_rts_fallthrough(input_stacks, z, render=render)
+    scales = np.array([ts.tforms[-1].scale for ts in rts.tilespecs])
+    return scales[np.all(scales != [1, 1], axis=1)]
+
+def do_get_z_scales_nopm(*args, **kwargs):
+    try:
+        return get_z_scales_nopm(*args, **kwargs)
+    except Exception:
+        return
+
+def get_scale_statistics_std(scales):
+    """get scale statistic for x and y"""
+    # throw out outliers
+    rel_scales = scales[np.all(np.abs(scales.mean(axis=0) - scales) < scales.std(axis=0) * 3, axis=1)]
+    return rel_scales.std(axis=0)
+
+def mad(data, axis=None, side=None):
+    '''Median Absolute Distance function with left and right side option.'''
+    def get_sided(d, side, axis=None):
+        # FIXME uses masked arrays for convenience, not for speed.
+        if side == "l":
+            m = d <= np.median(d, axis=axis)
+        elif side == "r":
+            m = d >= np.median(d, axis=axis)
+        elif side is None:
+            return d
+        else:
+            raise ValueError("ERROR: please enter 'l', 'r', or None")
+        return np.ma.masked_array(d, m)
+    d = get_sided(data, side, axis)
+    res = np.ma.median(np.absolute(d - np.ma.median(d, axis)), axis)
+    return (res.data if isinstance(res, np.ma.masked_array) else res)
+
+# TODO is mad robust enough to ignore outliers?
+def get_scale_statistics_mad(scales):
+    rel_scales = scales[np.all(np.abs(scales.mean(axis=0) - scales) < scales.std(axis=0) * 3, axis=1)]
+    return mad(rel_scales, axis=0)
+
+def get_scale_statistics_mad_sum(scales):
+    rel_scales = scales[np.all(np.abs(scales.mean(axis=0) - scales) < scales.std(axis=0) * 3, axis=1)]
+    return mad(rel_scales, axis=0, side="l") + mad(rel_scales, axis=0, side="r")
+
+class DetectDistortedMontagesModule(RenderModule):
+    default_output_schema = DetectDistortionParametersOutput
+    default_schema = DetectDistortionParameters
+
+    def run(self):
+        zs = sorted(list({
+            i for l in (renderapi.stack.get_z_values_for_stack(s, render=self.render) for s in self.args['input_stacks']) for i in l
+        }))
+        final_zs = list(set(zs).intersection(set(self.args['zValues'])))
+
+        # get scale factor for each section
+        with concurrent.futures.ProcessPoolExecutor(max_workers=self.args['num_threads']) as e:
+            fut_to_z = {
+                e.submit(
+                    do_get_z_scales_nopm,
+                    z, self.args['input_stacks'], self.render): z
+                for z in final_zs
+            }
+            z_to_scales = {fut_to_z[fut]: fut.result() for fut in concurrent.futures.as_completed(fut_to_z.keys())}
+
+        # check if any scale is None
```

Comment:
I don't think it makes sense to retry these serially -- I think this was more of a manual reprocessing thing in the original notebook.

Reply:
can you elaborate on this comment. I couldn't relate it to the code you have highlighted. The highlighted code is part of the parallel processing.